File: Mira/Models/MemoryStore.swift
// MemoryStore.swift
// Mira's long-term memory: explicit preferences + observed patterns with confidence scores.
//
// Explicit memories  → confidence 0.95 (user directly stated)
// Observed memories  → confidence 0.40–0.70 (inferred from behavior)
// Confidence decays  → −0.01/day without reinforcement, floor 0.10
// Reinforcement      → +0.08 when confirmed/re-stated, cap 1.0
// Conflict           → new value replaces old when confidence ≥ existing − 0.1

import Foundation

// MARK: - Model

struct Memory: Codable, Identifiable {
    let id:           UUID
    var key:          String       // snake_case, e.g. "preferred_browser"
    var value:        String       // human-readable value, e.g. "Safari"
    var category:     Category
    var source:       Source
    var confidence:   Double       // 0.0 – 1.0
    var notes:        String?      // optional supporting context
    let createdAt:    Date
    var updatedAt:    Date
    var accessCount:  Int

    enum Category: String, Codable, CaseIterable {
        case preference = "preference"
        case project    = "project"
        case person     = "person"
        case fact       = "fact"
        case goal       = "goal"

        var icon: String {
            switch self {
            case .preference: return "slider.horizontal.3"
            case .project:    return "folder.fill"
            case .person:     return "person.fill"
            case .fact:       return "info.circle.fill"
            case .goal:       return "target"
            }
        }
    }

    enum Source: String, Codable {
        case explicit   // user stated directly ("I prefer Safari")
        case observed   // inferred from repeated behavior
    }

    var confidenceTier: Tier {
        if confidence >= 0.8 { return .high }
        if confidence >= 0.5 { return .medium }
        return .low
    }

    enum Tier { case high, medium, low }
}

// MARK: - Store

@MainActor
final class MemoryStore: ObservableObject {

    static let shared = MemoryStore()

    @Published private(set) var memories: [Memory] = []

    // MARK: - Public API

    /// Upsert a memory. If a matching key already exists, updates value and boosts confidence.
    @discardableResult
    func upsert(key: String,
                value: String,
                category: Memory.Category = .fact,
                source: Memory.Source = .explicit,
                confidence: Double = 0.95,
                notes: String? = nil) -> Memory {
        let normKey = key.lowercased().trimmingCharacters(in: .whitespaces)

        if let idx = memories.firstIndex(where: { $0.key.lowercased() == normKey }) {
            memories[idx].value       = value
            memories[idx].notes       = notes ?? memories[idx].notes
            memories[idx].updatedAt   = Date()
            memories[idx].accessCount += 1
            memories[idx].confidence  = min(1.0, max(confidence,
                                                      memories[idx].confidence + 0.08))
            persist()
            return memories[idx]
        }

        let mem = Memory(id: UUID(), key: key, value: value,
                         category: category, source: source,
                         confidence: confidence, notes: notes,
                         createdAt: Date(), updatedAt: Date(), accessCount: 1)
        memories.append(mem)
        persist()
        return mem
    }

    /// Fuzzy search memories by key or value substring.
    func recall(query: String) -> [Memory] {
        let q = query.lowercased()
        return memories
            .filter { $0.key.lowercased().contains(q) || $0.value.lowercased().contains(q) }
            .sorted { $0.confidence > $1.confidence }
    }

    func memory(forKey key: String) -> Memory? {
        memories.first { $0.key.lowercased() == key.lowercased() }
    }

    func delete(id: UUID) {
        memories.removeAll { $0.id == id }
        persist()
    }

    func delete(key: String) {
        memories.removeAll { $0.key.lowercased() == key.lowercased() }
        persist()
    }

    func clear() {
        memories = []
        try? FileManager.default.removeItem(at: storeURL)
    }

    /// Bumps confidence when a memory is confirmed or referenced.
    func reinforce(key: String) {
        guard let idx = memories.firstIndex(where: { $0.key.lowercased() == key.lowercased() }) else { return }
        memories[idx].confidence  = min(1.0, memories[idx].confidence + 0.08)
        memories[idx].accessCount += 1
        memories[idx].updatedAt   = Date()
        persist()
    }

    /// Applies daily decay to all memories. Call once per app launch.
    func applyDecay() {
        let now        = Date()
        let daySeconds = 86_400.0
        var changed    = false
        for idx in memories.indices {
            let days  = now.timeIntervalSince(memories[idx].updatedAt) / daySeconds
            let decay = days * 0.01
            if decay > 0.001 {
                memories[idx].confidence = max(0.1, memories[idx].confidence - decay)
                changed = true
            }
        }
        if changed { persist() }
    }

    // MARK: - Prompt block (injected into session instructions)

    /// Returns the top memories formatted for the system prompt.
    /// Only includes memories with confidence ≥ 0.5 to avoid injecting uncertain guesses.
    func buildPromptBlock() -> String {
        let relevant = memories
            .filter { $0.confidence >= 0.5 }
            .sorted { $0.confidence > $1.confidence }
            .prefix(12)

        guard !relevant.isEmpty else { return "" }

        var lines = ["[What I know about you]"]
        for mem in relevant {
            let qualifier: String = {
                switch mem.confidenceTier {
                case .high:   return ""
                case .medium: return " (probably)"
                case .low:    return " (uncertain)"
                }
            }()
            lines.append("\(mem.key): \(mem.value)\(qualifier)")
        }
        return lines.joined(separator: "\n")
    }

    // MARK: - Persistence

    private init() {
        load()
        applyDecay()
    }

    private func load() {
        guard let data = try? Data(contentsOf: storeURL),
              let decoded = try? JSONDecoder().decode([Memory].self, from: data) else { return }
        memories = decoded
    }

    private func persist() {
        guard let data = try? JSONEncoder().encode(memories) else { return }
        try? data.write(to: storeURL, options: .atomic)
    }

    private var storeURL: URL {
        let support = FileManager.default.urls(for: .applicationSupportDirectory,
                                               in: .userDomainMask).first!
        let dir = support.appendingPathComponent("Mira", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("memory.json")
    }
}

File: windows/Mira.Windows.Core/Memory/MemoryStore.cs
using Mira.Windows.Core.Storage;
using Newtonsoft.Json;

namespace Mira.Windows.Core.Memory;

/// <summary>
/// The Windows equivalent of Mira/Models/MemoryStore.swift — Mira's long-term
/// memory: explicit preferences + observed patterns with confidence scores
/// that reinforce on restatement and decay without it. Purely local, JSON-
/// persisted (mirrors Mac's own local JSON file, just under
/// <c>%LocalAppData%\Mira\memory.json</c> instead of Application Support),
/// no server involvement at all — matching the confirmed absence of any
/// Supabase table for this feature.
/// </summary>
public sealed class MemoryStore
{
    public static MemoryStore Shared { get; } = new();

    private const string FileName = "memory.json";
    private const double ReinforceBoost = 0.08;
    private const double DailyDecayRate = 0.01;
    private const double DecayFloor = 0.10;
    private const double PromptBlockMinConfidence = 0.5;
    private const int PromptBlockMaxItems = 12;

    private readonly List<Memory> _memories = new();
    private readonly object _gate = new();

    public event Action? Changed;

    private MemoryStore()
    {
        Load();
        ApplyDecay(DateTimeOffset.UtcNow);
    }

    public IReadOnlyList<Memory> Memories { get { lock (_gate) return _memories.ToList(); } }

    /// <summary>Upsert a memory. If a matching key already exists, updates the value and reinforces confidence rather than overwriting it outright — mirrors <c>MemoryStore.upsert(key:value:...)</c>.</summary>
    public Memory Upsert(string key, string value, MemoryCategory category = MemoryCategory.Fact,
        MemorySource source = MemorySource.Explicit, double confidence = 0.95, string? notes = null)
    {
        var now = DateTimeOffset.UtcNow;
        var normKey = key.Trim().ToLowerInvariant();
        Memory result;
        lock (_gate)
        {
            var existing = _memories.FirstOrDefault(m => m.Key.ToLowerInvariant() == normKey);
            if (existing is not null)
            {
                existing.Value = value;
                existing.Notes = notes ?? existing.Notes;
                existing.UpdatedAt = now;
                existing.AccessCount++;
                existing.Confidence = ComputeReinforcedConfidence(existing.Confidence, confidence);
                result = existing;
            }
            else
            {
                result = new Memory
                {
                    Id = Guid.NewGuid(),
                    Key = key,
                    Value = value,
                    Category = category,
                    Source = source,
                    Confidence = confidence,
                    Notes = notes,
                    CreatedAt = now,
                    UpdatedAt = now,
                    AccessCount = 1,
                };
                _memories.Add(result);
            }
        }
        Persist();
        Changed?.Invoke();
        return result;
    }

    /// <summary>Fuzzy search by key or value substring, highest confidence first — mirrors <c>recall(query:)</c>.</summary>
    public IReadOnlyList<Memory> Recall(string query)
    {
        var q = query.ToLowerInvariant();
        lock (_gate)
        {
            return _memories
                .Where(m => m.Key.ToLowerInvariant().Contains(q) || m.Value.ToLowerInvariant().Contains(q))
                .OrderByDescending(m => m.Confidence)
                .ToList();
        }
    }

    public Memory? MemoryForKey(string key)
    {
        lock (_gate) return _memories.FirstOrDefault(m => string.Equals(m.Key, key, StringComparison.OrdinalIgnoreCase));
    }

    public void Delete(Guid id)
    {
        lock (_gate) _memories.RemoveAll(m => m.Id == id);
        Persist();
        Changed?.Invoke();
    }

    public void Delete(string key)
    {
        lock (_gate) _memories.RemoveAll(m => string.Equals(m.Key, key, StringComparison.OrdinalIgnoreCase));
        Persist();
        Changed?.Invoke();
    }

    public void Clear()
    {
        lock (_gate) _memories.Clear();
        try { File.Delete(LocalAppData.PathFor(FileName)); } catch { /* best-effort */ }
        Changed?.Invoke();
    }

    /// <summary>Bumps confidence when a memory is confirmed or referenced — mirrors <c>reinforce(key:)</c>.</summary>
    public void Reinforce(string key)
    {
        lock (_gate)
        {
            var existing = _memories.FirstOrDefault(m => string.Equals(m.Key, key, StringComparison.OrdinalIgnoreCase));
            if (existing is null) return;
            existing.Confidence = Math.Min(1.0, existing.Confidence + ReinforceBoost);
            existing.AccessCount++;
            existing.UpdatedAt = DateTimeOffset.UtcNow;
        }
        Persist();
        Changed?.Invoke();
    }

    /// <summary>Applies daily decay to every memory — mirrors <c>applyDecay()</c>, called once per launch (see constructor).</summary>
    public void ApplyDecay(DateTimeOffset now)
    {
        var changed = false;
        lock (_gate)
        {
            foreach (var m in _memories)
            {
                var days = (now - m.UpdatedAt).TotalDays;
                var decayed = ComputeDecayedConfidence(m.Confidence, days);
                if (Math.Abs(decayed - m.Confidence) > 0.0001)
                {
                    m.Confidence = decayed;
                    changed = true;
                }
            }
        }
        if (changed) Persist();
    }

    /// <summary>The top memories formatted for splicing into the chat system prompt — mirrors <c>buildPromptBlock()</c>, including its ≥0.5-confidence floor so uncertain guesses never get injected as fact.</summary>
    public string BuildPromptBlock()
    {
        List<Memory> relevant;
        lock (_gate)
        {
            relevant = _memories
                .Where(m => m.Confidence >= PromptBlockMinConfidence)
                .OrderByDescending(m => m.Confidence)
                .Take(PromptBlockMaxItems)
                .ToList();
        }
        return FormatPromptBlock(relevant);
    }

    /// <summary>Pure — mirrors <c>upsert</c>'s reinforcement math: <c>min(1.0, max(proposed, existing + 0.08))</c>.</summary>
    public static double ComputeReinforcedConfidence(double existingConfidence, double proposedConfidence) =>
        Math.Min(1.0, Math.Max(proposedConfidence, existingConfidence + ReinforceBoost));

    /// <summary>Pure — mirrors <c>applyDecay</c>'s per-memory math: −0.01/day, floor 0.10.</summary>
    public static double ComputeDecayedConfidence(double confidence, double daysSinceUpdate)
    {
        var decay = daysSinceUpdate * DailyDecayRate;
        return decay > 0.001 ? Math.Max(DecayFloor, confidence - decay) : confidence;
    }

    /// <summary>Pure — mirrors <c>buildPromptBlock</c>'s text formatting, including the tier-based qualifier suffix.</summary>
    public static string FormatPromptBlock(IReadOnlyList<Memory> relevant)
    {
        if (relevant.Count == 0) return "";
        var lines = new List<string> { "[What I know about you]" };
        foreach (var m in relevant)
        {
            var qualifier = m.ConfidenceTier switch
            {
                MemoryConfidenceTier.High => "",
                MemoryConfidenceTier.Medium => " (probably)",
                _ => " (uncertain)",
            };
            lines.Add($"{m.Key}: {m.Value}{qualifier}");
        }
        return string.Join("\n", lines);
    }

    private void Persist()
    {
        List<Memory> snapshot;
        lock (_gate) snapshot = _memories.ToList();
        try { File.WriteAllText(LocalAppData.PathFor(FileName), JsonConvert.SerializeObject(snapshot)); }
        catch { /* best-effort persistence */ }
    }

    private void Load()
    {
        try
        {
            var path = LocalAppData.PathFor(FileName);
            if (!File.Exists(path)) return;
            var list = JsonConvert.DeserializeObject<List<Memory>>(File.ReadAllText(path));
            if (list is not null) _memories.AddRange(list);
        }
        catch
        {
            // Corrupt/missing file -- start fresh rather than throwing on construction.
        }
    }
}
